### PR TITLE
Added simple RNN denoiser

### PR DESCRIPTION
**Pull request by mryndzionek**
This is a denoiser based on a RNN (Recurrent Neural Network). It's floating point, but scaled down, so that running on RP2040 is possible. The architecture is roughly:
rnn

For such a small (microscopic even) network the performance is impressive in my opinion. On validation data, on a PC this is what I get:
https://github.com/user-attachments/assets/bdc9cea8-9ebb-44e6-a2ec-07c79cfda10c
Here is a spectrum comparison for a sample from the validation set:
deno_spec
On RP2040 it's somewhat worse, but still better than anything I tried with spectral subtraction.

Only works in USB and LSB mode for now. Fine tuning might be needed via regularization.

### CODE REVIEW

**Comment by Tonytti** [comment]
Amazing!

Il giorno mer 19 nov 2025 alle 13:24 mryndzionek ***@***.***>
ha scritto:

**Comment by dawsonjon** [comment]
Very impressive! Do you have any more technical background on this? I found I needed quite a strong signal but that it gave very good results. I got occasional signal cutouts sometimes, as if the signal was close to a squelch threshold.

**Comment by mryndzionek** [comment]
> Very impressive! Do you have any more technical background on this? I found I needed quite a strong signal but that it gave very good results. I got occasional signal cutouts sometimes, as if the signal was close to a squelch threshold.

I'm still fine-tuning. Another PR will most likely land today. The tricky part is getting the right balance between sensitivity and noise rejection. Additionally I'm also thresholding on the average (sum) of the output gains. This threshold might be configurable. The input data are also AGCed, but not in a very sophisticated way.

Regarding technical background, similar approach was studied, but for much bigger networks. Then it's possible to e.g. remove background noise like traffic, engine sound, or unintelligible background chatter. I suspected that much smaller network should be able to handle non-stationary white noise and impulse noise and it turned out to be the case. First network I heard of using this approach was [RNNoise](https://jmvalin.ca/demo/rnnoise/). There is also [PercepNet](https://www.amazon.science/blog/how-amazon-chimes-challenge-winning-noise-cancellation-works). The only differences are:
 - much smaller network, only 253 parameters!
 - RNN layer simpler and faster than GRU
 - network is trained on synthetic data (20k 1s speech snippets generated using a speech synthesizer)

RP2350 has FPU, so bigger model might be feasible.

**Comment by mryndzionek** [comment]
You can turn off the artificial threshold by commenting out [this](https://github.com/dawsonjon/PicoRX/blob/d82eb625c07df390ac368206761912c293e95734/rnn_denoiser.cpp#L149) condition.
